### DISCUSSION TITLE
Change the spacing in horizontal axis as the user zooms in line graph

### DISCUSSION THREAD

**Discussion by JayantSharma777** [patrykandpatrick/vico#424] [Questions] [answered]
I want to change the spacing in between the grid lines horizontal axis as the user zooms in the graph

- [ ] `compose`
- [x] `views`
- [ ] `compose-m2`
- [ ] `compose-m3`
- [ ] `core` (select if none of the above apply)

<img width="675" alt="image" src="https://github.com/patrykandpatrick/vico/assets/75779183/960f3c75-4a1d-4de5-9a03-57513b47396f">
Now I want to change this spacing as the user zooms in the graph,so can you provide any callback function that gets triggered when the user zooms the graph,or implement this functionality .

Expected behavior
The spacing gets changed as the user zooms in/out the graph.

Android versions
13, vico version - 1.9.2

**Comment by JayantSharma777** [comment]
@patrickmichalik please can you give your views on this?

**Reply by patrickmichalik** [reply, marked as answer]
Hello! You can create a custom `AxisItemPlacer.Horizontal` implementation and customize the behavior as required. `AxisItemPlacer.Horizontal#getLabelValues` is called for every frame, and it can certainly return different lists of _x_ values depending on the zoom level. The default `AxisItemPlacer.Horizontal` implementation is relatively complex, but please don’t let that discourage you. It’s written to perform well in a variety of cases, including extreme ones. An implementation specific to a single use case will definitely be simpler. The documentation for the `AxisItemPlacer.Horizontal` functions describes what they are to return.

**Reply by JayantSharma777** [reply]
Thanks @patrickmichalik and again
A very great library to use in android .

**Reply by patrickmichalik** [reply]
You’re welcome! Thanks for the kind words.

**Reply by zell-mbc** [reply]
@JayantSharma777 did you end up implementing a custom item placer? I need the exact same dynamic behaviour and am not sure how to approach this. Compose in my case, but that shouldn't matter I guess.

**Reply by patrickmichalik** [reply]
@zell-mbc, [Vico 2.0.0 Alpha 7](https://github.com/patrykandpatrick/vico/releases/tag/v2.0.0-alpha.7) introduced dynamic `HorizontalAxis` label spacing, which may help. Please see the release notes for details and an example. Whether this will suffice depends on the specifics of the desired result. If you share some more details, I’ll be happy to suggest an approach.

**Reply by zell-mbc** [reply]
Thanks Patrick, I will check out Dynamic spacing asap, but let me explain my use case:
Let's use BloodPressure as an example, a user might record the values twice a day, sometimes the readings are off which means he does another measurement a few minutes later to confirm. So we are talking ~750 records in a year.
What I want to do is that upon opening the chart the full timeframe is visible, so upwards downwards trends can easily be identified. 
Following that, there may be certain areas of interest which the user wants to drill into via zooming.

From a technical point of view I think the way I want to solve this is that I identify the xAxis real estate of the phone and determine how many labels will fit nicely on the screen, let's say 6.
So at launch of the chart I would show 6 labels, YYYY-mmm = "2024 Jan", "2024 Mar"….
As the user zooms into the data more, he reaches the point where only 6 months are visible, that's when I would change to monthly labels and lines
Next level of granularity could be week or more likely mmm-dd, 05-01, 05-02,…
Given that there can be multiple measurements in a day, my highest zoom and finest granularity would be minutes so I can see the two readings collected in the morning and the one in the evening spread over the minutes of a day. Similar to the requirement @JayantSharma777 mentioned here https://github.com/patrykandpatrick/vico/discussions/384

All of this is theory now, and I may end up doing something slightly different, because after all one can also move the chart left and right, so maybe I would switch to finer granularity earlier, don't know yet, would have to play with this.

Edit: Just read up on  dynamic label spacing and tried it out really quick. Looks like that will work. Together with a brushed up bottomAxisValueFormatter I should be able to do what I want. I'll keep you posted.
But now I am off for the Beer gardens, it's May 1st after all :-)

**Reply by patrickmichalik** [reply]
Got it, @zell-mbc. Dynamic `HorizontalAxis` label spacing uses the smallest multiple of `spacing` for which no overlaps occur. What you’ve described, on the other hand, involves switching between several predetermined spacing values that aren’t consecutive multiples of a single number. Therefore, dynamic `HorizontalAxis` label spacing would be insufficient—you’d end up with unwanted intermediate granularity levels. The behavior in question can be obtained by means of a custom implementation of `HorizontalAxis.ItemPlacer` (or, in Vico 2.0.0 Alpha 28 and earlier, `AxisItemPlacer.Horizontal`). It looks like you’ve started a dedicated discussion concerning dynamically changing the label format—#891—so I’ll omit that here. I hope you had a good time in the beer gardens!
